File: scripts/intelligent_router.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
intelligent_router.py - The Brain and Heart of ZEN AI RAG

Multi-tier intelligent routing system combining:
- Tier 0: Semantic Cache (instant responses)
- Tier 1: Mini RAG (local knowledge)
- Tier 2: Traffic Controller (fast classifier)
- Tier 3: Smart Routing (cost-optimal LLM selection)
- Tier 4: Full Consensus (hard questions)

Based on 2024-2025 production research from OpenAI, Anthropic, Stanford, ETH Zurich.
Target: 90%+ queries < 200ms, 70%+ cost savings, 95%+ accuracy.
"""
import asyncio
import time
import logging
import json
from pathlib import Path
from typing import AsyncGenerator, Optional, Tuple, Dict
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum

logger = logging.getLogger("IntelligentRouter")

# Optional imports
try:
    from semantic_cache import SemanticCache, CacheConfig
    from mini_rag import MiniRAG, MiniRAGConfig
    CACHE_AVAILABLE = True
except ImportError:
    CACHE_AVAILABLE = False
    logger.warning("[Router] semantic_cache or mini_rag not available")


# =============================================================================
# Configuration
# =============================================================================
class RoutingTier(Enum):
    """Routing tier for tracking."""
    CACHE = "cache"
    MINI_RAG = "mini_rag"
    FAST_LLM = "fast_llm"
    POWERFUL_LLM = "powerful_llm"
    CONSENSUS = "consensus"
    ERROR = "error"


@dataclass
class RouterConfig:
    """Intelligent router configuration."""

    # Tier enablement
    enable_cache: bool = True
    enable_mini_rag: bool = True
    enable_traffic_controller: bool = True

    # Cache settings
    cache_dir: Path = field(default_factory=lambda: Path("cache"))

    # Mini RAG settings
    knowledge_file: Path = field(default_factory=lambda: Path("knowledge_base.json"))

    # Traffic controller settings
    traffic_controller_port: int = 8020
    traffic_controller_enabled: bool = True

    # Confidence thresholds
    mini_rag_confidence: float = 0.85
    fast_llm_confidence: float = 0.7
    consensus_threshold: float = 0.7

    # Cost tracking
    track_costs: bool = True
    cost_per_token_fast: float = 0.00001    # $0.01 per 1M tokens
    cost_per_token_powerful: float = 0.00003  # $0.03 per 1M tokens

    # Monitoring
    log_routing_decisions: bool = True
    save_routing_history: bool = True


@dataclass
class RoutingDecision:
    """Single routing decision record."""
    query: str
    query_hash: str
    tier: RoutingTier
    latency_ms: float
    cost_usd: float
    confidence: float
    timestamp: datetime = field(default_factory=datetime.now)
    success: bool = True
    error_msg: str = ""

    def to_dict(self) -> Dict:
        return {
            "query_hash": self.query_hash,
            "tier": self.tier.value,
            "latency_ms": self.latency_ms,
            "cost_usd": self.cost_usd,
            "confidence": self.confidence,
            "timestamp": self.timestamp.isoformat(),
            "success": self.success,
            "error_msg": self.error_msg
        }


@dataclass
class RouterStats:
    """Router statistics."""
    total_queries: int = 0
    tier_counts: Dict[str, int] = field(default_factory=dict)
    total_cost_usd: float = 0.0
    total_latency_ms: float = 0.0
    errors: int = 0

    def record(self, decision: RoutingDecision):
        """Record a routing decision."""
        self.total_queries += 1
        tier = decision.tier.value
        self.tier_counts[tier] = self.tier_counts.get(tier, 0) + 1
        self.total_cost_usd += decision.cost_usd
        self.total_latency_ms += decision.latency_ms
        if not decision.success:
            self.errors += 1

    def get_summary(self) -> Dict:
        """Get summary statistics."""
        if self.total_queries == 0:
            return {"total_queries": 0}

        return {
            "total_queries": self.total_queries,
            "avg_latency_ms": self.total_latency_ms / self.total_queries,
            "avg_cost_usd": self.total_cost_usd / self.total_queries,
            "total_cost_usd": self.total_cost_usd,
            "error_rate": f"{self.errors / self.total_queries * 100:.2f}%",
            "tier_distribution": {
                tier: f"{count / self.total_queries * 100:.1f}%"
                for tier, count in self.tier_counts.items()
            },
            "cost_savings_vs_all_powerful": self._calculate_savings()
        }

    def _calculate_savings(self) -> str:
        """Calculate cost savings vs always using powerful LLM."""
        # Assume powerful LLM would cost 3x more on average
        baseline_cost = self.total_queries * 0.0003  # $0.0003 per query
        if baseline_cost == 0:
            return "0%"
        savings = (baseline_cost - self.total_cost_usd) / baseline_cost * 100
        return f"{savings:.1f}%"


# =============================================================================
# Intelligent Router
# =============================================================================
class IntelligentRouter:
    """
    Multi-tier intelligent routing system.

    Flow:
    1. Check semantic cache (Tier 0)
    2. Check mini RAG (Tier 1)
    3. Classify with traffic controller (Tier 2)
    4. Route to appropriate LLM(s) (Tier 3/4)
    """

    def __init__(
        self,
        config: Optional[RouterConfig] = None,
        swarm_arbitrator: Optional[object] = None
    ):
        self.config = config or RouterConfig()
        self.swarm = swarm_arbitrator  # Reference to SwarmArbitrator

        # Initialize cache (Tier 0)
        self.cache: Optional[SemanticCache] = None
        if CACHE_AVAILABLE and self.config.enable_cache:
            try:
                cache_config = CacheConfig()
                self.cache = SemanticCache(cache_config, self.config.cache_dir)
                logger.info("[Router] Semantic cache initialized")
            except Exception as e:
                logger.warning(f"[Router] Failed to initialize cache: {e}")

        # Initialize mini RAG (Tier 1)
        self.mini_rag: Optional[MiniRAG] = None
        if CACHE_AVAILABLE and self.config.enable_mini_rag:
            try:
                mini_rag_config = MiniRAGConfig()
                self.mini_rag = MiniRAG(mini_rag_config, self.config.knowledge_file)
                logger.info("[Router] Mini RAG initialized")
            except Exception as e:
                logger.warning(f"[Router] Failed to initialize mini RAG: {e}")

        # Statistics
        self.stats = RouterStats()
        self.routing_history: List[RoutingDecision] = []

        logger.info("[Router] Intelligent router initialized")

    # =========================================================================
    # Main Routing Method
    # =========================================================================

    async def route(
        self,
        query: str,
        system_prompt: str = "You are a helpful AI assistant.",
        stream: bool = True
    ) -> AsyncGenerator[str, None]:
        """
        Main routing method - the brain of the system.

        Args:
            query: User query
            system_prompt: System prompt for LLM
            stream: Whether to stream response

        Yields:
            Response chunks
        """
        start_time = time.time()
        query_hash = self._hash_query(query)

        try:
            # TIER 0: Semantic Cache (instant)
            if self.cache:
                cache_result = self.cache.get(query)

                if cache_result:
                    answer, source, confidence = cache_result
                    latency_ms = (time.time() - start_time) * 1000

                    # Record decision
                    decision = RoutingDecision(
                        query=query[:100],
                        query_hash=query_hash,
                        tier=RoutingTier.CACHE,
                        latency_ms=latency_ms,
                        cost_usd=0.0,  # No cost for cache
                        confidence=confidence
                    )
                    self._record_decision(decision)

                    # Show cache indicator
                    yield f"⚡ **Instant (cached)** - {source}\n\n"
                    yield answer
                    return

            # TIER 1: Mini RAG (local knowledge)
            if self.mini_rag:
                mini_rag_result = self.mini_rag.search(query)

                if mini_rag_result:
                    answer, confidence, category = mini_rag_result

                    if confidence >= self.config.mini_rag_confidence:
                        latency_ms = (time.time() - start_time) * 1000

                        # Record decision
                        decision = RoutingDecision(
                            query=query[:100],
                            query_hash=query_hash,
                            tier=RoutingTier.MINI_RAG,
                            latency_ms=latency_ms,
                            cost_usd=0.0,  # No LLM cost
                            confidence=confidence
                        )
                        self._record_decision(decision)

                        # Cache this answer
                        if self.cache:
                            self.cache.put(
                                query, answer,
                                source="mini_rag",
                                confidence=confidence
                            )

                        # Show mini RAG indicator
                        yield f"📚 **Knowledge Base** ({category}, {confidence:.0%} confidence)\n\n"
                        yield answer
                        return

            # TIER 2-4: Route to LLM(s) via SwarmArbitrator
            if self.swarm:
                # Delegate to SwarmArbitrator which handles:
                # - Tier 2: Traffic controller classification
                # - Tier 3: Smart routing (fast vs powerful)
                # - Tier 4: Full consensus (if needed)

                tier = RoutingTier.FAST_LLM  # Will be updated by swarm
                response_text = ""

                async for chunk in self.swarm.get_consensus(
                    query,
                    system_prompt=system_prompt,
                    verbose=False
                ):
                    response_text += chunk
                    yield chunk

                # Estimate tier from swarm (simplified)
                # In production, SwarmArbitrator would report this
                if len(response_text) > 0:
                    latency_ms = (time.time() - start_time) * 1000

                    # Estimate cost (simplified)
                    tokens = len(response_text.split()) * 1.3  # Rough estimate
                    cost_usd = tokens * self.config.cost_per_token_fast

                    decision = RoutingDecision(
                        query=query[:100],
                        query_hash=query_hash,
                        tier=tier,
                        latency_ms=latency_ms,
                        cost_usd=cost_usd,
                        confidence=0.8  # Default
                    )
                    self._record_decision(decision)

                    # Cache high-quality LLM responses
                    if self.cache and len(response_text) > 20:
                        self.cache.put(
                            query, response_text,
                            source="llm",
                            confidence=0.8
                        )

            else:
                # No swarm available - error
                yield "❌ **Error:** No LLM backend available\n"
                decision = RoutingDecision(
                    query=query[:100],
                    query_hash=query_hash,
                    tier=RoutingTier.ERROR,
                    latency_ms=(time.time() - start_time) * 1000,
                    cost_usd=0.0,
                    confidence=0.0,
                    success=False,
                    error_msg="No swarm arbitrator"
                )
                self._record_decision(decision)

        except Exception as e:
            logger.error(f"[Router] Error routing query: {e}")
            yield f"❌ **Error:** {str(e)}\n"

            decision = RoutingDecision(
                query=query[:100],
                query_hash=query_hash,
                tier=RoutingTier.ERROR,
                latency_ms=(time.time() - start_time) * 1000,
                cost_usd=0.0,
                confidence=0.0,
                success=False,
                error_msg=str(e)
            )
            self._record_decision(decision)

    # =========================================================================
    # Helper Methods
    # =========================================================================

    def _hash_query(self, query: str) -> str:
        """Generate hash of query."""
        import hashlib
        normalized = ' '.join(query.lower().strip().split())
        return hashlib.sha256(normalized.encode()).hexdigest()[:16]

    def _record_decision(self, decision: RoutingDecision):
        """Record routing decision."""
        self.stats.record(decision)
        self.routing_history.append(decision)

        if self.config.log_routing_decisions:
            logger.info(
                f"[Router] {decision.tier.value.upper()}: "
                f"{decision.latency_ms:.1f}ms, "
                f"${decision.cost_usd:.6f}, "
                f"{decision.confidence:.0%} confidence"
            )

    def get_stats(self) -> Dict:
        """Get router statistics."""
        stats = self.stats.get_summary()

        # Add tier-specific stats
        if self.cache:
            stats["cache_stats"] = self.cache.get_stats()

        if self.mini_rag:
            stats["mini_rag_stats"] = self.mini_rag.get_stats()

        return stats

    def export_history(self, filepath: Path):
        """Export routing history to JSON."""
        try:
            data = {
                "summary": self.get_stats(),
                "history": [d.to_dict() for d in self.routing_history[-1000:]],  # Last 1000
                "timestamp": datetime.now().isoformat()
            }

            with open(filepath, 'w', encoding='utf-8') as f:
                json.dump(data, f, indent=2)

            logger.info(f"[Router] Exported history to {filepath}")

        except Exception as e:
            logger.error(f"[Router] Failed to export history: {e}")

    def print_performance_report(self):
        """Print detailed performance report."""
        stats = self.get_stats()

        print("\n" + "="*70)
        print("INTELLIGENT ROUTER PERFORMANCE REPORT")
        print("="*70)

        print(f"\n📊 Overall Statistics:")
        print(f"  Total Queries:    {stats['total_queries']}")
        print(f"  Avg Latency:      {stats.get('avg_latency_ms', 0):.1f}ms")
        print(f"  Avg Cost:         ${stats.get('avg_cost_usd', 0):.6f}")
        print(f"  Total Cost:       ${stats.get('total_cost_usd', 0):.4f}")
        print(f"  Error Rate:       {stats.get('error_rate', '0%')}")
        print(f"  Cost Savings:     {stats.get('cost_savings_vs_all_powerful', '0%')}")

        print(f"\n🎯 Tier Distribution:")
        for tier, percentage in stats.get('tier_distribution', {}).items():
            print(f"  {tier:20s}: {percentage}")

        if 'cache_stats' in stats:
            cache_stats = stats['cache_stats']
            print(f"\n⚡ Cache Performance:")
            print(f"  Hit Rate:         {cache_stats.get('hit_rate', '0%')}")
            print(f"  Total Entries:    {cache_stats.get('total_entries', 0)}")
            print(f"  Exact Matches:    {cache_stats.get('exact_matches', 0)}")
            print(f"  Semantic Matches: {cache_stats.get('semantic_matches', 0)}")

        if 'mini_rag_stats' in stats:
            rag_stats = stats['mini_rag_stats']
            print(f"\n📚 Mini RAG Performance:")
            print(f"  Hit Rate:         {rag_stats.get('hit_rate', '0%')}")
            print(f"  Total Entries:    {rag_stats.get('total_entries', 0)}")
            print(f"  High Conf Hits:   {rag_stats.get('high_confidence_hits', 0)}")

        print("\n" + "="*70)


# =============================================================================
# Factory Functions
# =============================================================================
def create_intelligent_router(
    swarm_arbitrator: Optional[object] = None,
    config: Optional[RouterConfig] = None
) -> IntelligentRouter:
    """Create intelligent router with default configuration."""
    return IntelligentRouter(config=config, swarm_arbitrator=swarm_arbitrator)


# =============================================================================
# Main (for testing)
# =============================================================================
if __name__ == "__main__":
    import logging
    logging.basicConfig(level=logging.INFO)

    # Create router (without swarm for testing cache/rag only)
    router = create_intelligent_router()

    # Test queries
    async def test_router():
        test_queries = [
            "How do I start the LLM?",
            "What is SWARM_ENABLED?",
            "How to launch the model?",  # Should hit cache
            "What is the meaning of life?",  # Should miss all tiers
        ]

        for query in test_queries:
            print(f"\n{'='*70}")
            print(f"Query: {query}")
            print(f"{'='*70}")

            async for chunk in router.route(query):
                print(chunk, end='', flush=True)

            print("\n")

        # Print report
        router.print_performance_report()

    # Run test
    asyncio.run(test_router())

File: rust_output/src/intelligent_router.rs
/// intelligent_router::py - The Brain and Heart of ZEN AI RAG
/// 
/// Multi-tier intelligent routing system combining:
/// - Tier 0: Semantic Cache (instant responses)
/// - Tier 1: Mini RAG (local knowledge)
/// - Tier 2: Traffic Controller (fast classifier)
/// - Tier 3: Smart Routing (cost-optimal LLM selection)
/// - Tier 4: Full Consensus (hard questions)
/// 
/// Based on 2024-2025 production research from OpenAI, Anthropic, Stanford, ETH Zurich.
/// Target: 90%+ queries < 200ms, 70%+ cost savings, 95%+ accuracy.

use anyhow::{Result, Context};
use serde::{Serialize, Deserialize};
use std::collections::HashMap;
use std::fs::File;
use std::io::{self, Read, Write};
use std::path::PathBuf;
use tokio;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

/// Routing tier for tracking.
#[derive(Debug, Clone)]
pub struct RoutingTier {
}

/// Intelligent router configuration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RouterConfig {
    pub enable_cache: bool,
    pub enable_mini_rag: bool,
    pub enable_traffic_controller: bool,
    pub cache_dir: PathBuf,
    pub knowledge_file: PathBuf,
    pub traffic_controller_port: i64,
    pub traffic_controller_enabled: bool,
    pub mini_rag_confidence: f64,
    pub fast_llm_confidence: f64,
    pub consensus_threshold: f64,
    pub track_costs: bool,
    pub cost_per_token_fast: f64,
    pub cost_per_token_powerful: f64,
    pub log_routing_decisions: bool,
    pub save_routing_history: bool,
}

/// Single routing decision record.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RoutingDecision {
    pub query: String,
    pub query_hash: String,
    pub tier: RoutingTier,
    pub latency_ms: f64,
    pub cost_usd: f64,
    pub confidence: f64,
    pub timestamp: datetime,
    pub success: bool,
    pub error_msg: String,
}

impl RoutingDecision {
    /// To dict.
    pub fn to_dict(&self) -> HashMap {
        // To dict.
        HashMap::from([("query_hash".to_string(), self.query_hash), ("tier".to_string(), self.tier.value), ("latency_ms".to_string(), self.latency_ms), ("cost_usd".to_string(), self.cost_usd), ("confidence".to_string(), self.confidence), ("timestamp".to_string(), self.timestamp.isoformat()), ("success".to_string(), self.success), ("error_msg".to_string(), self.error_msg)])
    }
}

/// Router statistics.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RouterStats {
    pub total_queries: i64,
    pub tier_counts: HashMap<String, i64>,
    pub total_cost_usd: f64,
    pub total_latency_ms: f64,
    pub errors: i64,
}

impl RouterStats {
    /// Record a routing decision.
    pub fn record(&mut self, decision: RoutingDecision) -> () {
        // Record a routing decision.
        self.total_queries += 1;
        let mut tier = decision.tier.value;
        self.tier_counts[tier] = (self.tier_counts.get(&tier).cloned().unwrap_or(0) + 1);
        self.total_cost_usd += decision.cost_usd;
        self.total_latency_ms += decision.latency_ms;
        if !decision.success {
            self.errors += 1;
        }
    }
    /// Get summary statistics.
    pub fn get_summary(&mut self) -> HashMap {
        // Get summary statistics.
        if self.total_queries == 0 {
            HashMap::from([("total_queries".to_string(), 0)])
        }
        HashMap::from([("total_queries".to_string(), self.total_queries), ("avg_latency_ms".to_string(), (self.total_latency_ms / self.total_queries)), ("avg_cost_usd".to_string(), (self.total_cost_usd / self.total_queries)), ("total_cost_usd".to_string(), self.total_cost_usd), ("error_rate".to_string(), format!("{:.2}%", ((self.errors / self.total_queries) * 100))), ("tier_distribution".to_string(), self.tier_counts.iter().iter().map(|(tier, count)| (tier, format!("{:.1}%", ((count / self.total_queries) * 100)))).collect::<HashMap<_, _>>()), ("cost_savings_vs_all_powerful".to_string(), self._calculate_savings())])
    }
    /// Calculate cost savings vs always using powerful LLM.
    pub fn _calculate_savings(&mut self) -> String {
        // Calculate cost savings vs always using powerful LLM.
        let mut baseline_cost = (self.total_queries * 0.0003_f64);
        if baseline_cost == 0 {
            "0%".to_string()
        }
        let mut savings = (((baseline_cost - self.total_cost_usd) / baseline_cost) * 100);
        format!("{:.1}%", savings)
    }
}

/// Multi-tier intelligent routing system.
/// 
/// Flow:
/// 1. Check semantic cache (Tier 0)
/// 2. Check mini RAG (Tier 1)
/// 3. Classify with traffic controller (Tier 2)
/// 4. Route to appropriate LLM(s) (Tier 3/4)
#[derive(Debug, Clone)]
pub struct IntelligentRouter {
    pub config: String,
    pub swarm: String,
    pub cache: Option<SemanticCache>,
    pub mini_rag: Option<MiniRAG>,
    pub stats: RouterStats,
    pub routing_history: Vec<RoutingDecision>,
}

impl IntelligentRouter {
    /// Initialize instance.
    pub fn new(config: Option<RouterConfig>, swarm_arbitrator: Option<object>) -> Self {
        Self {
            config: (config || RouterConfig()),
            swarm: swarm_arbitrator,
            cache: None,
            mini_rag: None,
            stats: RouterStats(),
            routing_history: Vec::new(),
        }
    }
}

/// Route part1 part 2.
pub fn _route_part1_part2(r#self: String, query: String) -> Result<()> {
    // Route part1 part 2.
    let _hash_query = |query| {
        // Generate hash of query.
        // TODO: import hashlib
        let mut normalized = query.to_lowercase().trim().to_string().split_whitespace().map(|s| s.to_string()).collect::<Vec<String>>().join(&" ".to_string());
        hashlib::sha256(normalized.as_bytes().to_vec()).hexdigest()[..16]
    };
    let _record_decision = |decision| {
        // Record routing decision.
        self.stats.record(decision);
        self.routing_history.push(decision);
        if self.config::log_routing_decisions {
            logger.info(format!("[Router] {}: {:.1}ms, ${:.6}, {:.0%} confidence", decision.tier.value.to_uppercase(), decision.latency_ms, decision.cost_usd, decision.confidence));
        }
    };
    let get_stats = || {
        // Get router statistics.
        let mut stats = self.stats.get_summary();
        if self.cache {
            stats["cache_stats".to_string()] = self.cache::get_stats();
        }
        if self.mini_rag {
            stats["mini_rag_stats".to_string()] = self.mini_rag.get_stats();
        }
        stats
    };
    let export_history = |filepath| {
        // Export routing history to JSON.
        // try:
        {
            let mut data = HashMap::from([("summary".to_string(), self.get_stats()), ("history".to_string(), self.routing_history[-1000..].iter().map(|d| d.to_dict()).collect::<Vec<_>>()), ("timestamp".to_string(), datetime::now().isoformat())]);
            let mut f = File::create(filepath)?;
            {
                json::dump(data, f, /* indent= */ 2);
            }
            logger.info(format!("[Router] Exported history to {}", filepath));
        }
        // except Exception as e:
    Ok(})
}

/// Route part1 part 3.
pub fn _route_part1_part3(r#self: String) -> () {
    // Route part1 part 3.
    let print_performance_report = || {
        // Print detailed performance report.
        let mut stats = self.get_stats();
        println!("{}", ("\n".to_string() + ("=".to_string() * 70)));
        println!("{}", "INTELLIGENT ROUTER PERFORMANCE REPORT".to_string());
        println!("{}", ("=".to_string() * 70));
        println!("\n📊 Overall Statistics:");
        println!("  Total Queries:    {}", stats["total_queries".to_string()]);
        println!("  Avg Latency:      {:.1}ms", stats.get(&"avg_latency_ms".to_string()).cloned().unwrap_or(0));
        println!("  Avg Cost:         ${:.6}", stats.get(&"avg_cost_usd".to_string()).cloned().unwrap_or(0));
        println!("  Total Cost:       ${:.4}", stats.get(&"total_cost_usd".to_string()).cloned().unwrap_or(0));
        println!("  Error Rate:       {}", stats.get(&"error_rate".to_string()).cloned().unwrap_or("0%".to_string()));
        println!("  Cost Savings:     {}", stats.get(&"cost_savings_vs_all_powerful".to_string()).cloned().unwrap_or("0%".to_string()));
        println!("\n🎯 Tier Distribution:");
        for (tier, percentage) in stats.get(&"tier_distribution".to_string()).cloned().unwrap_or(HashMap::new()).iter().iter() {
            println!("  {:20}: {}", tier, percentage);
            // pass
        }
        if stats.contains(&"cache_stats".to_string()) {
            let mut cache_stats = stats["cache_stats".to_string()];
            println!("\n⚡ Cache Performance:");
            println!("  Hit Rate:         {}", cache_stats.get(&"hit_rate".to_string()).cloned().unwrap_or("0%".to_string()));
            println!("  Total Entries:    {}", cache_stats.get(&"total_entries".to_string()).cloned().unwrap_or(0));
            println!("  Exact Matches:    {}", cache_stats.get(&"exact_matches".to_string()).cloned().unwrap_or(0));
            println!("  Semantic Matches: {}", cache_stats.get(&"semantic_matches".to_string()).cloned().unwrap_or(0));
        }
        if stats.contains(&"mini_rag_stats".to_string()) {
            let mut rag_stats = stats["mini_rag_stats".to_string()];
            println!("\n📚 Mini RAG Performance:");
            println!("  Hit Rate:         {}", rag_stats.get(&"hit_rate".to_string()).cloned().unwrap_or("0%".to_string()));
            println!("  Total Entries:    {}", rag_stats.get(&"total_entries".to_string()).cloned().unwrap_or(0));
            println!("  High Conf Hits:   {}", rag_stats.get(&"high_confidence_hits".to_string()).cloned().unwrap_or(0));
        }
        println!("{}", ("\n".to_string() + ("=".to_string() * 70)));
    };
}

/// Route part 1.
pub fn _route_part1(r#self: String, query: String) -> Result<()> {
    // Route part 1.
    let mut decision = RoutingDecision(/* query= */ query[..100], /* query_hash= */ query_hash, /* tier= */ RoutingTier.ERROR, /* latency_ms= */ ((std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64() - start_time) * 1000), /* cost_usd= */ 0.0_f64, /* confidence= */ 0.0_f64, /* success= */ false, /* error_msg= */ e.to_string());
    self._record_decision(decision);
    let route = |query, system_prompt, stream| {
        // Main routing method - the brain of the system.
        // 
        // Args:
        // query: User query
        // system_prompt: System prompt for LLM
        // stream: Whether to stream response
        // 
        // Yields:
        // Response chunks
        let mut start_time = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64();
        let mut query_hash = self._hash_query(query);
        // try:
        {
            if self.cache {
                let mut cache_result = self.cache::get(&query).cloned();
                if cache_result {
                    let (mut answer, mut source, mut confidence) = cache_result;
                    let mut latency_ms = ((std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64() - start_time) * 1000);
                    let mut decision = RoutingDecision(/* query= */ query[..100], /* query_hash= */ query_hash, /* tier= */ RoutingTier.CACHE, /* latency_ms= */ latency_ms, /* cost_usd= */ 0.0_f64, /* confidence= */ confidence);
                    self._record_decision(decision);
                    /* yield format!("⚡ **Instant (cached)** - {}\n\n", source) */;
                    /* yield answer */;
                    return;
                }
            }
            if self.mini_rag {
                let mut mini_rag_result = self.mini_rag.search(query);
                if mini_rag_result {
                    let (mut answer, mut confidence, mut category) = mini_rag_result;
                    if confidence >= self.config::mini_rag_confidence {
                        let mut latency_ms = ((std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64() - start_time) * 1000);
                        let mut decision = RoutingDecision(/* query= */ query[..100], /* query_hash= */ query_hash, /* tier= */ RoutingTier.MINI_RAG, /* latency_ms= */ latency_ms, /* cost_usd= */ 0.0_f64, /* confidence= */ confidence);
                        self._record_decision(decision);
                        if self.cache {
                            self.cache::put(query, answer, /* source= */ "mini_rag".to_string(), /* confidence= */ confidence);
                        }
                        /* yield format!("📚 **Knowledge Base** ({}, {:.0%} confidence)\n\n", category, confidence) */;
                        /* yield answer */;
                        return;
                    }
                }
            }
            if self.swarm {
                let mut tier = RoutingTier.FAST_LLM;
                let mut response_text = "".to_string();
                // async for
                while let Some(chunk) = self.swarm.get_consensus(query, /* system_prompt= */ system_prompt, /* verbose= */ false).next().await {
                    response_text += chunk;
                    /* yield chunk */;
                }
                if response_text.len() > 0 {
                    let mut latency_ms = ((std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64() - start_time) * 1000);
                    let mut tokens = (response_text.split_whitespace().map(|s| s.to_string()).collect::<Vec<String>>().len() * 1.3_f64);
                    let mut cost_usd = (tokens * self.config::cost_per_token_fast);
                    let mut decision = RoutingDecision(/* query= */ query[..100], /* query_hash= */ query_hash, /* tier= */ tier, /* latency_ms= */ latency_ms, /* cost_usd= */ cost_usd, /* confidence= */ 0.8_f64);
                    self._record_decision(decision);
                    if (self.cache && response_text.len() > 20) {
                        self.cache::put(query, response_text, /* source= */ "llm".to_string(), /* confidence= */ 0.8_f64);
                    }
                }
            } else {
                /* yield "❌ **Error:** No LLM backend available\n".to_string() */;
                let mut decision = RoutingDecision(/* query= */ query[..100], /* query_hash= */ query_hash, /* tier= */ RoutingTier.ERROR, /* latency_ms= */ ((std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64() - start_time) * 1000), /* cost_usd= */ 0.0_f64, /* confidence= */ 0.0_f64, /* success= */ false, /* error_msg= */ "No swarm arbitrator".to_string());
                self._record_decision(decision);
            }
        }
        // except Exception as e:
        _route_part1(self, query);
    };
    _route_part1_part2(self, query);
    Ok(_route_part1_part3(self))
}

/// Create intelligent router with default configuration.
pub fn create_intelligent_router(swarm_arbitrator: Option<object>, config: Option<RouterConfig>) -> IntelligentRouter {
    // Create intelligent router with default configuration.
    IntelligentRouter(/* config= */ config, /* swarm_arbitrator= */ swarm_arbitrator)
}
